lib/client.ts

```diff
@@ -0,0 +1,361 @@
+import { ServiceClientRequestOptions, ServiceClientResponse, request } from './request'
+import * as url from 'url'
+
+const retry = require('retry')
+const CircuitBreaker = require('circuit-breaker-js')
+
+/**
+ * A request filter may introduce one or both functions in this interface. For more information
+ * regarding request filters, refer to the readme of this project.
+ */
+export interface ServiceClientRequestFilter {
+  /**
+   * This callback is called before the requests is done.
+   * You can short-circuit the request by both returning
+   * a ServiceClient.Response object which is helpful for
+   * implementing caching or mocking. You could also
+   * fail the request by throwing an Error.
+   * @throws {Error}
+   */
+  request?: (requestOptions: ServiceClientRequestOptions) => Promise<ServiceClientResponse| ServiceClientRequestOptions>
```

Comment:
Formatting, a space before `|`. Maybe we need to add this rule to TSLint

Reply:
standard-ts?